clash-cores/src/Clash/Cores/Xilinx/DcFifo.hs

```diff
@@ -0,0 +1,271 @@
+
+Furthermore, the FIFO is configured as follows:
+
+* /Native/ interface type
+* /Independent clocks block RAM/ implementation
+* 2 synchronization stages
+* /Standard FIFO/ read mode
+* /No/ output registers
+* Reset pin /enabled/, reset synchronization /disabled/, safety circuit
+/disabled/
+
+    (Note: the GUI will still say /Asynchronous Reset/, grayed out. This might be
+    confusing: the resets are actually synchronous.)
+
+* Full flag reset value: /@0 (False)@/
+
+    (Note: this is actually not configurable for synchronous resets, it is
+    always @False@.)
+
+* No /Dout reset value/
+* Optional overflow flag: 'dcOverflow'
+* Optional underflow flag: 'dcUnderflow'
+* No other configurable status flags (no almost full, almost empty, write
+acknowledge, valid, or programmable full\/empty flags)
+* Optional write data count: 'dcWriteDataCount' (only full width is supported)
+* Optional read data count: 'dcReadDataCount' (only full width is supported)
+
+(The order of these items corresponds to Vivado's /Customize IP/ GUI dialog in
+Vivado 2022.1.)
+
+== Reset sequencing
+
+The FIFO only supports synchronous resets ('Clash.Signal.Synchronous'). If the
+read or write domain has an asynchronous reset signal, simulating the FIFO or
+generating HDL for it will throw an error.
+
+The FIFO does not need to be reset; it is possible to use the FIFO with its
+reset inputs permanently deasserted. However, resetting the FIFO will flush it,
+so resets do have a use. It is necessary to generate a proper reset sequence;
+see the product guide for details. Read and write side need to be reset
+together, both for at least one clock cycle. During the time where either or
+both of the resets are asserted, do not perform write or read operations to
+avoid unexpected behavior.
+
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ViewPatterns #-}
+
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+module Clash.Cores.Xilinx.DcFifo
+  ( -- * Instantiating IP
+    dcFifo
+  , FifoOut(..)
+
+    -- * Customizing IP
+  , DcConfig(..)
+  , defConfig
+
+    -- * Helper type aliases
+  , Full
+  , Empty
+  , DataCount
+  ) where
+
+import Clash.Explicit.Prelude
+import Clash.Signal.Internal (Clock (..), Signal (..))
+import Data.Maybe (isJust)
+import qualified Data.Sequence as Seq
+import Data.String.Interpolate (i)
+import Numeric.Natural (Natural)
+
+import Clash.Annotations.Primitive (Primitive (InlineYamlPrimitive))
+
+import Clash.Cores.Xilinx.DcFifo.Internal.Instances ()
+import Clash.Cores.Xilinx.DcFifo.Internal.Types
+
+-- | Default config. Read\/write data counts are enabled but over-\/underflow
+-- are not.
+--
+-- Examples using a type-level argument to specify the depth:
+--
+-- @
+-- fifo = dcFifo (defConfig @5)
+-- @
+-- @
+-- fifo = dcFifo @5 defConfig
+-- @
+-- @
+-- fifo = dcFifo (defConfig @5){dcReadDataCount=False}
+-- @
+-- @
+-- fifo = dcFifo @5 defConfig{dcReadDataCount=False}
+-- @
```

Comment:
You could make this a doctest. The REPL doesn't care if you redefine variable names - and the code would be typechecked on CI which is a plus.